Documentation/PrettyPrinter.md

```diff
@@ -0,0 +1,260 @@
+
+# Maximum line length of 10
+Output =
+"""
+one
+two
+  three
+"""
+
+Tokens = ["one", break(offset=2), open(C, 0), "two", break, "three", close]
+
+# Maximum line length of 10
+Output =
+"""
+one
+  two
+  three
+"""
+```
+
+The open token of a group is assigned the total size of the group as its length.
+Open tokens must always be paired with a *close* token.
+
+```
+Tokens = ["one", break(1, 2), open(C, 2), "two", break(1, 2), "three", close]
+Lengths = [3, 11, 10, 3, 1, 5, 0]
+```
+
+#### Close
+
+The *close* tokens indicate the end of a group, and they have a length of zero.
+They must always be paired with an *open* token.
+
+#### Newline
+
+The *newline* tokens behave much the same way as *break* tokens, except that
+they always produce a line break. They can be assigned an offset, in the same
+way as a break. They can also be given an integer number of line breaks to
+produce.
+
+These tokens are given a length equal to the maximum allowed line width. The
+reason for this is to indicate that any enclosing groups are too large to fit on
+a single line.
+
+```
+# Assume maximum line length of 50
+# break(size)
+Tokens = ["one", break(1), "two", break(1), open, "three", newline, "four", close]
+Lengths = [3, 4, 3, 60, 59, 5, 50, 4, 0]
+```
+
+#### Reset
+
+Reset tokens are used to reset the state created by break tokens if needed, and
+are rarely used. A primary use-case is to prevent an entire group from moving to
+a new line, but you still want the group to break internally. Reset tokens have
+a length of zero.
+
+A reset token makes whatever follows it behave as if it was at the beginning of
+the line.
+
+```
+Tokens = ["one", break(1), "two", reset]
+Lengths = [3, 4, 3, 0]
+
+# Normal breaking behavior of a consistent group
+Tokens = ["one", break(1), open(C, 2), "two", break(1), "three", break(1), "four", close]
+Output =
+"""
+one
+  two
+  three
+  four
+"""
+
+# Breaking behavior of a consistent group with a reset token
+Tokens = ["one", break(1), reset, open(C, 2), "two", break(1), "three", break(1), "four", close]
+Output =
+"""
+one two
+  three
+  four
+"""
+```
+
+#### Comment
+
+Comment tokens represent Swift source comments, and they come in four types:
+`line`, `docLine`, `block`, and `docBlock`. Their length is equal to the number
+of characters needed to print them, including whitespace and delimiters. Line
+comments produce one comment token per line. If other comment types span
+multiple lines, their content is represented as a single comment token.
+
+```
+# Line comment
+// comment 1
+// comment 2
+Tokens = [line(" comment 1"), newline, line(" comment 2")]
+
+/// Doc comment 1
+/// Second line
+Tokens = [docLine(" Doc comment 1\n Second line")]
+
+/* Block comment
+   Second line */
+Tokens = [block(" Block comment\n   Second Line ")]
+
+/** Doc Block comment
+  * Second line **/
+Tokens = [docBlock(" Doc Block comment\n  * Second line *")]
+```
+
+### Token Generation
```

Comment:
Unless you have some more detailed write-up of this planned in a future PR, it would be useful to `visit(TokenSyntax)` is where the text of tokens (and comments) is handled and that the remaining visitors should just be used to add groups/breaks.

Reply:
Yes, that is a good idea. I can expand that section in the next PR.

Reply:
SG, thanks!